src/polydatum/dal.py

```diff
@@ -1,20 +1,239 @@
 from __future__ import absolute_import
 from contextlib import contextmanager
-from .context import DataAccessContext
 from polydatum.errors import AlreadyExistsException
 from polydatum.util import is_generator
 from .resources import ResourceManager
 from .context import _ctx_stack
+from typing import Callable, Tuple, Dict, Optional, Any
+from functools import partial, update_wrapper
+from polydatum.context import DataAccessContext
+
+
+class PathSegment:
+    """
+    This structure represents a segment of an attribute path.
+
+    For example, in this code:
+
+        dal.my_service.sub_service.other_service.method()
+
+    For the `dal` object, the entire attribute path would be:
+
+        my_service.sub_service.other_service.method
+
+    Which may be made up of PathSegments like:
+
+        PathSegment(name="my_service"), PathSegment(name="sub_service"), etc.
+    """
+    name = None
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta = None
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+    def __eq__(self, other):
```

Comment:
What happened to `@dataclass`? Doesn't that handle this?

Reply:
Not available in 3.5

Reply:
> Python 3.5 has reached end-of-life. Python 3.5.10 is the final release of 3.5.

https://www.python.org/downloads/release/python-3510/

We can drop it. Embrace the future. What are we using, 3.7? 3.9?